Sources/NIO/BSDSocketAPI.swift

```diff
@@ -365,3 +460,533 @@ extension NIOBSDSocket.Option {
             NIOBSDSocket.Option(rawValue: SO_TIMESTAMP)
 }
 #endif
+
+extension NIOBSDSocket {
+    @inline(never)
+    internal static func accept(socket s: NIOBSDSocket.Handle,
+                                address addr: UnsafeMutablePointer<sockaddr>?,
+                                address_len addrlen: UnsafeMutablePointer<socklen_t>?)
+            throws -> NIOBSDSocket.Handle? {
+        #if os(iOS) || os(macOS) || os(tvOS) || os(watchOS)
+            guard case let .processed(fd) = try syscall(blocking: true, { () -> NIOBSDSocket.Handle in
+                let fd = cAccept(s, addr, addrlen)
+                guard fd > 0 else {
+                    return fd
+                }
+
+                do {
+                    try Posix.fcntl(descriptor: fd, command: F_SETNOSIGPIPE, value: 1)
+                } catch {
+                    _ = cClose(fd)  // don't care about failure here
+                    throw error
+                }
+
+                return fd
+            }) else {
+                return nil
+            }
+
+            return fd
+        #elseif os(Windows)
+              let socket: NIOBSDSocket.Handle = WinSDK.accept(s, addr, addrlen)
+              if socket == WinSDK.INVALID_SOCKET {
+                  throw IOError(winsock: WSAGetLastError(), reason: "accept")
+              }
+              return socket
+        #else
+            guard case let .processed(fd) = try syscall(blocking: true, {
+                cAccept(s, addr, addrlen)
+            }) else {
+                return nil
+            }
+
+            return fd
+        #endif
+    }
+
+    @inline(never)
+    internal static func bind(socket s: NIOBSDSocket.Handle,
+                              address addr: UnsafePointer<sockaddr>,
+                              address_len namelen: socklen_t) throws {
+        #if os(Windows)
```

Comment:
I really struggle to read this `#if os` code. Couldn't we just rename `System.swift` to `SystemUNIX.swift` and make a Windows version that defines exactly the same ones without

```
#if os(Windows)
    complicatedCode()
#else if os(...) 
    moreComplicatedCode()
#endif
```

NIO is in active development and we have to maintain all this so I'd really appreciate if we could try to make the code easy to read and follow.

Reply:
I think `ThreadPosix.swift` & `ThreadWindows.swift` are good examples how the code can be still easy to read and yet portable. It's very clear which file you have to open, there's no performance overhead, and because of the (unused) protocol we still have clarity on what has to be implemented with what exact types.

Reply:
Doesn't that exactly violate your idea of you having one place to find the code? :)  I think that it would be better to introduce the indirection in a follow up change much like you did with the threading.  I think that it is going to be really odd having `BSDSocketAPIUnix.swift` and `BSDSocketWindows.swift` though, but thats just a minor point.  IMO this current structure does make it easier to compare the behaviour which is valuable.

Reply:
I personally find code that is full of `#if os(...)` impossible to read. I would just go to `BSDSocketAPIUnix.swift` and that gives me all the code. If something's only not working on Windows, I'd go to `BSDSocketAPIWindows.swift`.

Sure, for the porting effort it's very valuable to see both behaviours at once but after the porting is done I think it's more important that the code is readable and I really struggle hard with lots of `#if os(...)`. I'm fine with them if all is super trivial and restricted to very few files. But the code sockets API full of `#if os(...)`es would be a touch pill for me to swallow.

CC @Lukasa / @glbrntt / @PeterAdams-A / @Davidde94 how do you feel about lots of `#if os(...)` interspersed in code?

Reply:
I'm a +1 on not having lots of `#if os(...)`: splitting into unix/windows specific files is my preferred option here.

Reply:
In general I think the more `#if`s then the harder the code is to read and maintain. 2 files would work, but I'm sure it would also be possible to elegantly group code within one file but with far fewer directives.

Reply:
@Davidde94 what did you have in mind?  I can try to regroup things for clarity.

Reply:
Hey @weissi @Davidde94 can you revisit this? I'd like to unblock this work as it will also unblock #1473 and subsequent work.

Reply:
I don't think we can take lots of `#if os(...)` in non-trivial code.

Reply:
In my opinion, this file does the exact opposite of what I think makes maintainable code. It clumps together different implementations for the same interface.

I think we should create the interface and then have two files implementing it. One for Windows, one for Posix.

I'm also not happy with the syscalls being taken out of `System.swift`. `System.swift` is our (poorly named) place where we define all Posix syscalls. We have many times done refactorings to exactly how we call into those syscalls and I think it's crucial that they stay in one place which is `System.swift`. I'm happy renaming it `PosixSystem.swift` or so.